api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -28,27 +26,66 @@ option (xds.annotations.v3.file_status).work_in_progress = true;
 // .. code-block:: yaml
 //
 //    authentication_header: "X-API-KEY"
-//    keys:
-//      inline_string: |-
-//        clientID1:apiKey1
-//        clientID2:apiKey2
+//    credentials:
+//      entries:
+//      - api_key: 09876abcdefg
+//        client_id: user
+//
+// If multiple fields of ``authentication_header``, ``authentication_query``, and ``authentication_cookie``
+// are set. Then filter will try to fetch the API key from the following fields in order:
+//
+// 1. ``authentication_header``
+// 2. ``authentication_query``
+// 3. ``authentication_cookie``
 //
-message APIKeyAuth {
-  // keys used to authenticate the client.
-  // It should be a map of clientID to apiKey.
-  // The clientID serves solely for identification purposes and isn't used for authentication.
-  config.core.v3.DataSource keys = 1 [(udpa.annotations.sensitive) = true];
+// If none of the fields are set, then the filter will reject the request.
+message ApiKeyAuth {
+  message Credential {
+    // The value of the unique API key.
+    string api_key = 1 [(validate.rules).string = {min_len: 1}];
+
+    // The unique id or identity that used to identify the client or consumer.
+    string client_id = 2 [(validate.rules).string = {min_len: 1}];
+  }
+
+  message Credentials {
+    repeated Credential entries = 1;
+  }
+
+  // Api credentials used to authenticate the clients.
+  Credentials credentials = 1 [(udpa.annotations.sensitive) = true];
 
   // The header name to fetch the key.
   // If multiple values are present in the given header, the filter rejects the request.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_header = 2;
+  string authentication_header = 2
+      [(validate.rules).string =
+           {max_len: 1024 well_known_regex: HTTP_HEADER_NAME strict: false ignore_empty: true}];
 
   // The query parameter name to fetch the key.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_query = 3;
+  string authentication_query = 3 [(validate.rules).string = {max_len: 1024}];
 
   // The cookie name to fetch the key.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_cookie = 4;
+  string authentication_cookie = 4
+      [(validate.rules).string =
+           {max_len: 1024 well_known_regex: HTTP_HEADER_NAME strict: false ignore_empty: true}];
+}
+
+// API key auth configuration of per route or per virtual host or per route configuration.
+message ApiKeyAuthPerScope {
```

Comment:
Because this per route featue of api key auth is pretty simple and almost doesn't introduce too much additional code, I will prefer to place it here directly.